File: PANDO_backend/pipelines/matching/scorers.py
"""
Phase 2.3 feature matching: one scorer per match type, each comparing a UserDNA field against
a Property field and returning a float in [0, 1] (0 = no match, 1 = perfect match).

Shared convention: when the relevant UserDNA field is unset (the user never stated that
preference), the scorer returns NEUTRAL_SCORE rather than 0 or 1 — an unstated preference must
not be treated as either a strong match or a penalty. This mirrors the missing-value philosophy
from Phase 1.2/2.2 (None means "not stated," not "no").
"""

from pipelines.cleaning.text_cleaning import normalize_key
from pipelines.schemas.property_schema import Property
from pipelines.schemas.user_dna_schema import Purpose, RentalYieldPreference, TransportPreference, UserDNA

NEUTRAL_SCORE = 0.75

# Rental yield bands (annual %, typical Dubai market ranges) used to translate a property's
# numeric rental_yield into the same High/Medium/Low vocabulary as RentalYieldPreference.
_YIELD_HIGH_THRESHOLD = 7.0
_YIELD_MEDIUM_THRESHOLD = 5.0

# How far over budget a price can be before Budget Match bottoms out at 0. A property exactly
# at budget.max scores 1.0; one at (1 + _OVERAGE_TOLERANCE) x budget.max scores 0.0.
_OVERAGE_TOLERANCE = 0.5


def _property_price(property_: Property) -> float | None:
    return property_.price if property_.listing_purpose.value == "Sale" else property_.rent_price


def score_budget_match(user_dna: UserDNA, property_: Property) -> float:
    if user_dna.budget is None or (user_dna.budget.min is None and user_dna.budget.max is None):
        return NEUTRAL_SCORE

    price = _property_price(property_)
    if price is None:
        return NEUTRAL_SCORE

    low = user_dna.budget.min
    high = user_dna.budget.max

    if low is not None and price < low:
        deficit_ratio = (low - price) / low if low > 0 else 1.0
        return max(0.0, 1.0 - deficit_ratio)

    if high is not None and price > high:
        overage_ratio = (price - high) / high if high > 0 else 1.0
        return max(0.0, 1.0 - overage_ratio / _OVERAGE_TOLERANCE)

    return 1.0


def score_location_match(user_dna: UserDNA, property_: Property) -> float:
    if not user_dna.locations:
        return NEUTRAL_SCORE
    wanted = {normalize_key(loc) for loc in user_dna.locations}
    return 1.0 if normalize_key(property_.location) in wanted else 0.0


def score_property_type_match(user_dna: UserDNA, property_: Property) -> float:
    if not user_dna.property_types:
        return NEUTRAL_SCORE
    wanted = {normalize_key(t) for t in user_dna.property_types}
    return 1.0 if normalize_key(property_.property_type) in wanted else 0.0


def score_bedroom_match(user_dna: UserDNA, property_: Property) -> float:
    if user_dna.bedrooms is None:
        return NEUTRAL_SCORE
    diff = abs(property_.bedrooms - user_dna.bedrooms)
    if diff == 0:
        return 1.0
    return max(0.0, 1.0 - 0.25 * diff)


_PURPOSE_TO_PROPERTY_PURPOSE = {
    Purpose.INVESTMENT: "Investment",
    Purpose.END_USE: "End use",
    Purpose.RENTAL: "Rental",
}


def score_purpose_match(user_dna: UserDNA, property_: Property) -> float:
    if user_dna.purpose is None:
        return NEUTRAL_SCORE
    wanted = _PURPOSE_TO_PROPERTY_PURPOSE[user_dna.purpose]
    property_purposes = {p.value for p in property_.property_purpose}
    return 1.0 if wanted in property_purposes else 0.0


def score_amenity_match(user_dna: UserDNA, property_: Property) -> float:
    if not user_dna.amenities:
        return NEUTRAL_SCORE
    wanted = {normalize_key(a) for a in user_dna.amenities}
    available = {normalize_key(a) for a in property_.amenities}
    if not wanted:
        return NEUTRAL_SCORE
    matched = wanted & available
    return len(matched) / len(wanted)


def score_area_match(user_dna: UserDNA, property_: Property) -> float:
    """UserDNA has no stated area/sqft preference field (see Phase 2.3 decision) — always
    neutral until such a field exists to compare against."""
    return NEUTRAL_SCORE


def score_transport_match(user_dna: UserDNA, property_: Property) -> float:
    if user_dna.transport_preference is None or user_dna.transport_preference == TransportPreference.NO_PREFERENCE:
        return NEUTRAL_SCORE

    metro_distance = property_.nearby_facilities.metro_distance_km
    if metro_distance is None:
        return NEUTRAL_SCORE

    if metro_distance <= 1.0:
        return 1.0
    if metro_distance >= 5.0:
        return 0.0
    return 1.0 - (metro_distance - 1.0) / 4.0


def _yield_band(rental_yield: float) -> RentalYieldPreference:
    if rental_yield >= _YIELD_HIGH_THRESHOLD:
        return RentalYieldPreference.HIGH
    if rental_yield >= _YIELD_MEDIUM_THRESHOLD:
        return RentalYieldPreference.MEDIUM
    return RentalYieldPreference.LOW


_YIELD_BAND_ORDER = [RentalYieldPreference.LOW, RentalYieldPreference.MEDIUM, RentalYieldPreference.HIGH]


def score_investment_match(user_dna: UserDNA, property_: Property) -> float:
    if (
        user_dna.rental_yield_preference is None
        or user_dna.rental_yield_preference == RentalYieldPreference.NO_PREFERENCE
    ):
        return NEUTRAL_SCORE

    if property_.rental_yield is None:
        return NEUTRAL_SCORE

    wanted_band = user_dna.rental_yield_preference
    actual_band = _yield_band(property_.rental_yield)

    if wanted_band == actual_band:
        return 1.0

    band_distance = abs(_YIELD_BAND_ORDER.index(wanted_band) - _YIELD_BAND_ORDER.index(actual_band))
    return max(0.0, 1.0 - 0.5 * band_distance)


ALL_SCORERS = {
    "budget_match": score_budget_match,
    "location_match": score_location_match,
    "property_type_match": score_property_type_match,
    "bedroom_match": score_bedroom_match,
    "purpose_match": score_purpose_match,
    "amenity_match": score_amenity_match,
    "area_match": score_area_match,
    "transport_match": score_transport_match,
    "investment_match": score_investment_match,
}


def score_all(user_dna: UserDNA, property_: Property) -> dict[str, float]:
    return {name: scorer(user_dna, property_) for name, scorer in ALL_SCORERS.items()}

File: lib/buyerDna/scorers.js
/**
 * Phase 2.3 Feature Matching Scorers.
 * Ported from Python pipelines/matching/scorers.py
 */

import { normalizeKey, getCanonicalLocationKey } from './extractor.js';
import { Purpose, RentalYieldPreference, TransportPreference } from './schema.js';

export const NEUTRAL_SCORE = 0.75;
const YIELD_HIGH_THRESHOLD = 7.0;
const YIELD_MEDIUM_THRESHOLD = 5.0;
const OVERAGE_TOLERANCE = 0.5;

function getPropertyPrice(property) {
  if (!property) return null;
  const listingPurpose = property.listing_purpose || property.listingPurpose || 'Sale';
  if (typeof listingPurpose === 'string' && listingPurpose.toLowerCase().includes('rent')) {
    return typeof property.rent_price === 'number' ? property.rent_price : property.rentPrice || property.price || null;
  }
  return typeof property.price === 'number' ? property.price : null;
}

export function scoreBudgetMatch(userDna, property) {
  if (!userDna || !userDna.budget || (userDna.budget.min === null && userDna.budget.max === null)) {
    return NEUTRAL_SCORE;
  }

  const price = getPropertyPrice(property);
  if (price === null || price === undefined) {
    return NEUTRAL_SCORE;
  }

  const low = userDna.budget.min;
  const high = userDna.budget.max;

  if (low !== null && price < low) {
    const deficitRatio = low > 0 ? (low - price) / low : 1.0;
    return Math.max(0.0, 1.0 - deficitRatio);
  }

  if (high !== null && price > high) {
    const overageRatio = high > 0 ? (price - high) / high : 1.0;
    return Math.max(0.0, 1.0 - overageRatio / OVERAGE_TOLERANCE);
  }

  return 1.0;
}

export function scoreLocationMatch(userDna, property) {
  if (!userDna || !Array.isArray(userDna.locations) || userDna.locations.length === 0) {
    return NEUTRAL_SCORE;
  }

  const wanted = new Set(userDna.locations.map(getCanonicalLocationKey).filter(Boolean));
  const propertyLoc = getCanonicalLocationKey(property ? property.location || property.community || '' : '');

  return wanted.has(propertyLoc) ? 1.0 : 0.0;
}

export function scorePropertyTypeMatch(userDna, property) {
  if (!userDna || !Array.isArray(userDna.property_types) || userDna.property_types.length === 0) {
    return NEUTRAL_SCORE;
  }

  const wanted = new Set(
    userDna.property_types.map((t) => {
      let k = normalizeKey(t);
      return k.endsWith('s') && k !== 'bus' ? k.replace(/s$/, '') : k;
    }).filter(Boolean)
  );

  let propertyType = normalizeKey(property ? property.property_type || property.propertyType || property.category || '' : '');
  if (propertyType.endsWith('s') && propertyType !== 'bus') {
    propertyType = propertyType.replace(/s$/, '');
  }

  return wanted.has(propertyType) ? 1.0 : 0.0;
}

export function scoreBedroomMatch(userDna, property) {
  if (!userDna || userDna.bedrooms === null || userDna.bedrooms === undefined) {
    return NEUTRAL_SCORE;
  }

  const propertyBedrooms = property && typeof property.bedrooms === 'number' ? property.bedrooms : 0;
  const diff = Math.abs(propertyBedrooms - userDna.bedrooms);

  if (diff === 0) return 1.0;
  return Math.max(0.0, 1.0 - 0.25 * diff);
}

const PURPOSE_TO_PROPERTY_PURPOSE = {
  [Purpose.INVESTMENT]: 'Investment',
  [Purpose.END_USE]: 'End use',
  [Purpose.RENTAL]: 'Rental',
};

export function scorePurposeMatch(userDna, property) {
  if (!userDna || !userDna.purpose) {
    return NEUTRAL_SCORE;
  }

  const wanted = PURPOSE_TO_PROPERTY_PURPOSE[userDna.purpose];
  let propertyPurposes = [];

  if (property) {
    if (Array.isArray(property.property_purpose)) {
      propertyPurposes = property.property_purpose.map((p) => (typeof p === 'object' ? p.value || p.label : p));
    } else if (Array.isArray(property.propertyPurpose)) {
      propertyPurposes = property.propertyPurpose;
    } else if (property.purchasingGoal || property.investmentInsight) {
      propertyPurposes = ['Investment', 'End use', 'Rental'];
    } else {
      const listingPurp = String(property.listing_purpose || property.listingPurpose || property.purpose || '').toLowerCase();
      if (listingPurp.includes('rent')) {
        propertyPurposes = ['Rental', 'Investment'];
      } else if (listingPurp.includes('sale') || listingPurp.includes('buy')) {
        propertyPurposes = ['Investment', 'End use'];
      } else {
        propertyPurposes = ['Investment', 'End use', 'Rental'];
      }
    }
  }

  return propertyPurposes.some((p) => String(p).toLowerCase() === String(wanted).toLowerCase()) ? 1.0 : 0.0;
}

export function scoreAmenityMatch(userDna, property) {
  if (!userDna || !Array.isArray(userDna.amenities) || userDna.amenities.length === 0) {
    return NEUTRAL_SCORE;
  }

  const wanted = new Set(userDna.amenities.map(normalizeKey).filter(Boolean));
  if (wanted.size === 0) return NEUTRAL_SCORE;

  const availableList = property && Array.isArray(property.amenities) ? property.amenities : [];
  const available = new Set(availableList.map(normalizeKey).filter(Boolean));

  let matchedCount = 0;
  for (const item of wanted) {
    if (available.has(item)) matchedCount++;
  }

  return matchedCount / wanted.size;
}

export function scoreAreaMatch(userDna, property) {
  return NEUTRAL_SCORE;
}

export function scoreTransportMatch(userDna, property) {
  if (
    !userDna ||
    !userDna.transport_preference ||
    userDna.transport_preference === TransportPreference.NO_PREFERENCE
  ) {
    return NEUTRAL_SCORE;
  }

  let metroDistance = null;
  if (property) {
    if (property.nearby_facilities && typeof property.nearby_facilities.metro_distance_km === 'number') {
      metroDistance = property.nearby_facilities.metro_distance_km;
    } else if (property.nearbyFacilities && typeof property.nearbyFacilities.metroDistanceKm === 'number') {
      metroDistance = property.nearbyFacilities.metroDistanceKm;
    } else if (typeof property.metroDistanceKm === 'number') {
      metroDistance = property.metroDistanceKm;
    }
  }

  if (metroDistance === null || metroDistance === undefined) {
    return NEUTRAL_SCORE;
  }

  if (metroDistance <= 1.0) return 1.0;
  if (metroDistance >= 5.0) return 0.0;
  return 1.0 - (metroDistance - 1.0) / 4.0;
}

function getYieldBand(yieldValue) {
  if (yieldValue >= YIELD_HIGH_THRESHOLD) return RentalYieldPreference.HIGH;
  if (yieldValue >= YIELD_MEDIUM_THRESHOLD) return RentalYieldPreference.MEDIUM;
  return RentalYieldPreference.LOW;
}

const YIELD_BAND_ORDER = [
  RentalYieldPreference.LOW,
  RentalYieldPreference.MEDIUM,
  RentalYieldPreference.HIGH,
];

export function scoreInvestmentMatch(userDna, property) {
  if (
    !userDna ||
    !userDna.rental_yield_preference ||
    userDna.rental_yield_preference === RentalYieldPreference.NO_PREFERENCE
  ) {
    return NEUTRAL_SCORE;
  }

  const yieldValue = property
    ? typeof property.rental_yield === 'number'
      ? property.rental_yield
      : typeof property.yield === 'number'
      ? property.yield
      : null
    : null;

  if (yieldValue === null || yieldValue === undefined) {
    return NEUTRAL_SCORE;
  }

  const wantedBand = userDna.rental_yield_preference;
  const actualBand = getYieldBand(yieldValue);

  if (wantedBand === actualBand) return 1.0;

  const wantedIdx = YIELD_BAND_ORDER.indexOf(wantedBand);
  const actualIdx = YIELD_BAND_ORDER.indexOf(actualBand);
  const bandDistance = Math.abs(wantedIdx - actualIdx);

  return Math.max(0.0, 1.0 - 0.5 * bandDistance);
}

export const ALL_SCORERS = {
  budget_match: scoreBudgetMatch,
  location_match: scoreLocationMatch,
  property_type_match: scorePropertyTypeMatch,
  bedroom_match: scoreBedroomMatch,
  purpose_match: scorePurposeMatch,
  amenity_match: scoreAmenityMatch,
  area_match: scoreAreaMatch,
  transport_match: scoreTransportMatch,
  investment_match: scoreInvestmentMatch,
};

export function scoreAll(userDna, property) {
  const result = {};
  for (const [name, scorer] of Object.entries(ALL_SCORERS)) {
    result[name] = scorer(userDna, property);
  }
  return result;
}
